### PR TITLE
Added simple schedule request solver

### PR DESCRIPTION
**Pull request by marmor157**

### CODE REVIEW

File: lib/hunger_games/assign_schedule_worker.ex [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+
+  alias HungerGames.{
+    AssignedSchedules,
+    SimpleScheduleRequestSolver,
+    Schedules,
+    Schedules.Schedule,
+    Repo
+  }
+
+  def enqueue_schedule(%Schedule{id: id, registration_end_date: date}) do
```
Comment (Qizot) [inline lib/hunger_games/assign_schedule_worker.ex:19]: `@spec` :P

File: lib/hunger_games/assign_schedule_worker.ex [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+        }
+      end)
+
+    Multi.new()
+    |> Multi.run(:insert_assigned_schedules, fn _repo, _changes ->
+      result =
+        assigned_schedules
+        |> Enum.map(&AssignedSchedules.create_assigned_schedule/1)
+
+      case Enum.all?(result, &match?({:ok, _}, &1)) do
```
Comment (Qizot) [inline lib/hunger_games/assign_schedule_worker.ex:54]: you could pipe the result above into the `all` and then do the case as a pipe as well

File: lib/hunger_games/model/assigned_schedules.ex [hunk 3]
```diff
@@ -49,7 +52,35 @@ defmodule HungerGames.AssignedSchedules do
...
-  def create_assigned_schedule(attrs \\ %{}) do
+  def create_assigned_schedule(attrs \\ %{})
+
+  def create_assigned_schedule(%{classes: classes} = attrs) when is_list(classes) do
+    Multi.new()
+    |> Multi.insert(
+      :insert_assigned_schedule,
+      AssignedSchedule.changeset(%AssignedSchedule{}, attrs)
+    )
+    |> Multi.insert_all(:insert_all, "assigned_schedule_classes", fn changes ->
```
Comment (Qizot) [inline lib/hunger_games/model/assigned_schedules.ex:63]: Why string and not atom?

File: lib/hunger_games/schedule_requests_solver/schedule_request_solver.ex [hunk 4]
```diff
@@ -0,0 +1,14 @@
+defmodule HungerGames.ScheduleRequestSolver do
+  alias HungerGames.{Schedules.Schedule}
+
+  @type class :: %{
```
Comment (Qizot) [inline lib/hunger_games/schedule_requests_solver/schedule_request_solver.ex]: why not a struct?

File: lib/hunger_games/schedule_requests_solver/simple_schedule_request_solver.ex [hunk 5]
```diff
@@ -0,0 +1,119 @@
+defmodule HungerGames.SimpleScheduleRequestSolver do
+  @moduledoc """
+  This is a simplest possible ScheduleRequestSolver.
+
+  It is based on time of requests and does not feature collision detection
+  """
+  @behaviour HungerGames.ScheduleRequestSolver
```
Comment (Qizot) [inline lib/hunger_games/schedule_requests_solver/simple_schedule_request_solver.ex:7]: You use the behaviour but I can't see any `@impl`

File: lib/hunger_games/schedule_requests_solver/simple_schedule_request_solver.ex [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+    |> Enum.group_by(fn %{name: name, type: type} ->
+      name <> Atom.to_string(type)
+    end)
+  end
+
+  defp resolve_requests(%{requests: requests} = args) when length(requests) == 0, do: args
+
+  defp resolve_requests(%{requests: requests, classes: classes}) do
+    requests
+    |> Enum.reduce(%{classes: classes}, fn request, %{classes: classes} ->
```
Comment (Qizot) [inline lib/hunger_games/schedule_requests_solver/simple_schedule_request_solver.ex]: do you need the accumulator to be a map? You can wrap it after the reduce phase. It takes more space this way

File: lib/hunger_games/schedule_requests_solver/simple_schedule_request_solver.ex [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+      %{
+        classes: resolve_request(%{classes: classes, request: request})
+      }
+    end)
+  end
+
+  defp resolve_request(%{classes: classes, request: request}) do
```
Comment (Qizot) [inline lib/hunger_games/schedule_requests_solver/simple_schedule_request_solver.ex:79]: WIthout any docs it is hard to follow with those maps juggling.

**Review by Qizot — CHANGES_REQUESTED** [review]

File: lib/hunger_games/assign_schedule_worker.ex (hunk 2 above)
Comment (marmor157) [reply lib/hunger_games/assign_schedule_worker.ex:54]: That way I cannot return result in return tuple

File: lib/hunger_games/model/assigned_schedules.ex (hunk 3 above)
Comment (marmor157) [reply lib/hunger_games/model/assigned_schedules.ex:63]: Because it does not work with atom since while defining a schema a string was provided

File: lib/hunger_games/schedule_requests_solver/schedule_request_solver_class.ex [hunk 8]
```diff
@@ -0,0 +1,18 @@
+defmodule HungerGames.ScheduleRequestSolver.Class do
+  @enforce_keys [:id, :rrule, :size_limit, :name, :type]
+  defstruct id: "",
+            rrule: "",
+            size_limit: 0,
+            current_students: [],
+            name: "",
+            type: :exercises
+
+  @type t :: %__MODULE__{
```
Comment (Qizot) [inline lib/hunger_games/schedule_requests_solver/schedule_request_solver_class.ex:2]: first type spec then defstruct

File: lib/hunger_games/schedule_requests_solver/simple_schedule_request_solver.ex [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+    ScheduleRequestSolver.Class,
+    Schedules.Schedule
+  }
+
+  @impl true
+  def solve_schedule(%Schedule{classes: classes, requests: requests}) do
+    classes =
+      classes
+      |> Enum.map(&Map.from_struct(&1))
+      |> Enum.map(&struct(Class, &1))
```
Comment (Qizot) [inline lib/hunger_games/schedule_requests_solver/simple_schedule_request_solver.ex:21]: what was your intention here? struct to map and back to struct?

**Review by Qizot — CHANGES_REQUESTED** [review]

File: lib/hunger_games/schedule_requests_solver/simple_schedule_request_solver.ex (hunk 9 above)
Comment (marmor157) [reply lib/hunger_games/schedule_requests_solver/simple_schedule_request_solver.ex:21]: From database Class, to ScheduleRequestSolver Class to limit to filter out unnecessary fields and add some others

**Review by Qizot — APPROVED** [review]
